### PR TITLE
feat(payment_methods): Filter payment methods based on pm client secret 

### PR DESCRIPTION
**Pull request by Sarthak1799**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This PR builds upon -  https://github.com/juspay/hyperswitch/pull/4134
After generating a `pm client_secret`, `/ist_payment_methods` can be altered for vaulting purposes.
For saved payment methods, if `pm client_secret` is passed in merchant payment method list, the payment methods will be filtered on a config `saved_payment_methods`.
Only the payment methods listed in the config `saved_payment_methods` would be displayed as a part of "Vaulting eligible payment methods". 
The customer can then select one of the listed payment_methods, after which the SDK will hit the `/save` API for vaulting it.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [x] This PR modifies application configuration/environment variables

## How did you test it?

1. Create a customer
```
curl --location --request POST 'http://localhost:8080/customers' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_y5XDvbHKvRI6K2dAhTgXvNgQykJiqLXeYtSysCQzPGBGz3RZoUPjPcfD1LW8teYO' \
--data-raw '{
    "email": "guest@example.com",
    "name": "John Doe1",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "First customer",
    "address": {
        "line1": "1467",
        "line2": "Harrison Street",
        "line3": "Harrison Street",
        "city": "San Fransico",
        "state": "California",
        "zip": "94122",
        "country": "US",
        "first_name": "John",
        "last_name": "Doe"
    },
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}'
```
2. Create a payment method
```
curl --location --request POST 'http://localhost:8080/payment_methods' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_63yd7Y0i2CvRcJYKfXcIE8wAva0b5YMtKHr6oTpcnggVmWqsK0Q6CvoKB1WiXoTz' \
--data-raw '{
    "customer_id": "cus_RHnmMuG8mKhICwigyUeP"
}'
```
Response - 
```
{
    "merchant_id": "sarthak",
    "customer_id": "cus_RHnmMuG8mKhICwigyUeP",
    "payment_method_id": "pm_8b24uUXEoZIBTcUIPRBD",
    "payment_method": null,
    "payment_method_type": null,
    "card": null,
    "recurring_enabled": false,
    "installment_payment_enabled": false,
    "payment_experience": null,
    "metadata": null,
    "created": null,
    "last_used_at": null,
    "client_secret": "pm_8b24uUXEoZIBTcUIPRBD_secret_FyFrpYRzEU0NH0zbQJXC"
}
```
3. Pass pm client secret in merchant pm list - 
```
curl --location --request GET 'http://localhost:8080/account/payment_methods?client_secret=pm_8b24uUXEoZIBTcUIPRBD_secret_FyFrpYRzEU0NH0zbQJXC' \
--header 'Accept: application/json' \
--header 'api-key: pk_dev_ee1bf1a13355408da5f22cbee38ffca4'
```
Response - 
image

4. Save the payment method with client_secret
```
curl --location --request POST 'http://localhost:8080/payment_methods/pm_8b24uUXEoZIBTcUIPRBD/save' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: pk_dev_660e47ec672f43849d6b0403fcaf76ac' \
--data-raw '{
    "payment_method": "card",
    "client_secret": "pm_8b24uUXEoZIBTcUIPRBD_secret_FyFrpYRzEU0NH0zbQJXC",
    "customer_id": "cus_RHnmMuG8mKhICwigyUeP",
    "payment_method_data": {
        "card": {
            "card_number": "4242424242424242",
            "card_exp_month": "10",
            "card_exp_year": "2026",
            "card_holder_name": "joseph"
        }
    }
}'
```
Response - 
```
{
    "merchant_id": "sarthak",
    "customer_id": "cus_RHnmMuG8mKhICwigyUeP",
    "payment_method_id": "pm_DC74n7ocq8SKGvNVai0M",
    "payment_method": "card",
    "payment_method_type": null,
    "card": {
        "scheme": null,
        "issuer_country": null,
        "last4_digits": "4242",
        "expiry_month": "10",
        "expiry_year": "2026",
        "card_token": null,
        "card_holder_name": "joseph",
        "card_fingerprint": null,
        "nick_name": null,
        "card_network": null,
        "card_isin": "424242",
        "card_issuer": null,
        "card_type": null,
        "saved_to_locker": true
    },
    "recurring_enabled": false,
    "installment_payment_enabled": false,
    "payment_experience": [
        "redirect_to_url"
    ],
    "metadata": null,
    "created": "2024-04-03T13:02:00.926Z",
    "last_used_at": "2024-04-03T13:02:00.926Z",
    "client_secret": "pm_8b24uUXEoZIBTcUIPRBD_secret_FyFrpYRzEU0NH0zbQJXC"
}
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Review by Chethan-rao — CHANGES_REQUESTED** [review]

**Review by Chethan-rao — APPROVED** [review]

File: config/development.toml [hunk 1]
```diff
@@ -615,3 +615,6 @@ payment_attempts = "hyperswitch-payment-attempt-events"
 payment_intents = "hyperswitch-payment-intent-events"
 refunds = "hyperswitch-refund-events"
 disputes = "hyperswitch-dispute-events"
+
+[allowed_payment_methods_for_pm_client_secret]
```
Comment (vspecky) [inline config/development.toml:619]: We should go for a more appropriate name, along with namespacing if necessary. Something like
```
[saved_payment_methods]
eligible_via_sdk = ["card"]
```
for example. Feel free to choose a name as appropriate.

File: crates/router/src/core/payment_methods/cards.rs [hunk 2]
```diff
@@ -1691,12 +1691,39 @@ pub async fn list_payment_methods(
...
-    )
-    .await?;
+    let payment_intent = if let Some(cs) = &req.client_secret {
+        if cs.starts_with("pm_") {
+            let pm_vec = cs.split("_secret").collect::<Vec<&str>>();
+            let pm_id = pm_vec
+                .first()
+                .ok_or(errors::ApiErrorResponse::MissingRequiredField {
+                    field_name: "client_secret",
+                })?;
```
Comment (vspecky) [inline crates/router/src/core/payment_methods/cards.rs:1701]: This logic can be split into a helper function.

File: crates/router/src/core/payment_methods/cards.rs [hunk 3]
```diff
@@ -1691,12 +1691,39 @@ pub async fn list_payment_methods(
...
+                .await
+                .change_context(errors::ApiErrorResponse::PaymentMethodNotFound)
+                .attach_printable("Unable to find payment method")?;
+
+            let client_secret_expired =
+                authenticate_pm_client_secret_and_check_expiry(cs, &payment_method)?;
+            if client_secret_expired {
+                return Err((errors::ApiErrorResponse::ClientSecretExpired).into());
+            }
+
```
Comment (vspecky) [inline crates/router/src/core/payment_methods/cards.rs:1714]: Are we sure it is okay to not return early here?

**Review by vspecky — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payment_methods/cards.rs (hunk 3 above)
Comment (Sarthak1799) [reply crates/router/src/core/payment_methods/cards.rs:1714]: Could you elaborate a bit on this?

**Review by vspecky — DISMISSED** [review]

**Review by Chethan-rao — DISMISSED** [review]

**Review by Chethan-rao — APPROVED** [review]

**Review by Chethan-rao — APPROVED** [review]

**Review by vspecky — APPROVED** [review]

**Review by vspecky — APPROVED** [review]

**Review by Aprabhat19 — APPROVED** [review]

**Pull request merged**